[Sofa.Config][MSVC] Change SIMD cmake option and flags (#2652)
sofa-framework/sofa: SofaKernel/modules/Sofa.Config/CMakeLists.txt

option(SOFA_NO_UPDATE_BBOX "Compile Sofa with the SOFA_NO_UPDATE_BBOX macro defined." OFF)
option(SOFA_DUMP_VISITOR_INFO "Compile Sofa with the SOFA_DUMP_VISITOR_INFO macro defined." OFF)
if(MSVC)
    option(SOFA_VECTORIZE "(deprecated) Enable the use of SSE2 instructions by the compiler (Only available for MSVC)." OFF)

>     option(SOFA_ENABLE_SIMD "Enable the use of SIMD instructions by the compiler (AVX/AVX2 for msvc, march=native for gcc/clang)." OFF)

Reviewer: ```suggestion
    option(SOFA_ENABLE_SIMD "Enable the use of SIMD instructions by the compiler (AVX/AVX2 for msvc)." OFF)
```
`SOFA_ENABLE_SIMD` is available for MSVC only, right?

Author: I have the same for linux/gcc/clang in an other branch (and mac/clang as well)

>     option(SOFA_ENABLE_FAST_MATH "Enable floating-point model to fast (theorically faster but can bring unexpected results/bugs)." OFF)

Reviewer: ```suggestion
    option(SOFA_ENABLE_FAST_MATH "Enable floating-point model to fast (theoretically faster but can bring unexpected results/bugs)." OFF)
```

Author: Ha, I searched for that one : https://en.wiktionary.org/wiki/theorically 🤠

            # remove lots of warnings (the Wall of "normal" clang seems different of Wall of clang-cl)
            list(APPEND SOFACONFIG_COMPILE_OPTIONS -Wno-c++98-compat -Wno-c++98-compat-pedantic -Wno-double-promotion -Wno-old-style-cast -Wno-reserved-id-macro -Wno-language-extension-token -Wno-dllexport-explicit-instantiation-decl -Wno-nonportable-system-include-path -Wno-zero-as-null-pointer-constant -Wno-documentation)
            # optimization flags (not sure if necessary..)

>             list(APPEND SOFACONFIG_COMPILE_OPTIONS_RELEASE -march=native)

Reviewer: `march=native` means it is not possible to package SOFA for another processor than the one used to compile it.
I know Clang compilation on Windows is experimental. I just wanted to clarify it.

Author: I agree, we should put some warnings  here  and there.
But actually it is the same with the compilation with AVX2 on msvc, it wont work on systems without.
So those options should not be enabled when shipping the binaries.

    # Increase Number of Sections in .Obj file
    list(APPEND SOFACONFIG_COMPILE_OPTIONS "/bigobj")
    if(SOFA_VECTORIZE)

>         message("SOFA_VECTORIZE cmake option is deprecated and has been split between into SOFA_ENABLE_SIMD and SOFA_ENABLE_FAST_MATH\n. These 2 options will be set.")

Reviewer 2: ```suggestion
        message("SOFA_VECTORIZE option is deprecated and has been split between SOFA_ENABLE_SIMD and SOFA_ENABLE_FAST_MATH. These 2 options will be enabled.")
```

        set(SOFA_ENABLE_SIMD ON)
        set(SOFA_ENABLE_FAST_MATH ON)
    endif()
    # SIMD flags
    if(SOFA_ENABLE_SIMD)
        include(find_avx)
        MSVC_CHECK_FOR_AVX() # search for availability of AVX2 (otherwise set to AVX)
        if (HAVE_AVX2_EXTENSIONS)
            message("Your CPU supports AVX2 SIMD, set ${AVX_FLAGS} flags.")
        elseif(HAVE_AVX_EXTENSIONS)

>             message("Your CPU supports only AVX SIMD, set ${AVX_FLAGS} flags.")

Reviewer 2: ```suggestion
            message("Your CPU supports AVX2 SIMD. Using flags: ${AVX_FLAGS}")
        elseif(HAVE_AVX_EXTENSIONS)
            message("Your CPU supports only AVX SIMD. Using flags: ${AVX_FLAGS}")
```
